src/CloudNative.CloudEvents/CloudEventFormatter.cs

```diff
@@ -0,0 +1,117 @@
+        /// <param name="extensions">The extension attributes to use when populating the CloudEvent. May be null.</param>
+        /// <returns>The decoded CloudEvent.</returns>
+        public virtual CloudEvent DecodeStructuredModeMessage(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes)
+        {
+            var bytes = BinaryDataUtilities.ToByteArray(data);
+            return DecodeStructuredModeMessage(bytes, contentType, extensionAttributes);
+        }
+
+        /// <summary>
+        /// Decodes a CloudEvent from a structured-mode message body, represented as a stream. The default implementation asynchronously copies the
+        /// content of the stream into a byte array before passing it to <see cref="DecodeStructuredModeMessage(byte[], ContentType, IEnumerable{CloudEventAttribute})"/>
+        /// but this can be overridden by event formatters that can decode a stream more efficiently.
+        /// </summary>
+        /// <param name="data">The data within the message body. Must not be null.</param>
+        /// <param name="contentType">The content type of the message, or null if no content type is known.
+        /// Typically this is a content type with a media type of "application/cloudevents"; the additional
+        /// information such as the charset parameter may be needed in order to decode the data.</param>
+        /// <param name="extensions">The extension attributes to use when populating the CloudEvent. May be null.</param>
+        /// <returns>The CloudEvent derived from the structured data.</returns>
+        public virtual async Task<CloudEvent> DecodeStructuredModeMessageAsync(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes)
+        {
+            var bytes = await BinaryDataUtilities.ToByteArrayAsync(data).ConfigureAwait(false);
+            return DecodeStructuredModeMessage(bytes, contentType, extensionAttributes);
+        }
+
+        /// <summary>
+        /// Decodes a CloudEvent from a structured-mode message body, represented as a byte array.
+        /// </summary>
+        /// <param name="data">The data within the message body. Must not be null.</param>
+        /// <param name="contentType">The content type of the message, or null if no content type is known.
+        /// Typically this is a content type with a media type of "application/cloudevents"; the additional
+        /// information such as the charset parameter may be needed in order to decode the data.</param>
+        /// <param name="extensions">The extension attributes to use when populating the CloudEvent. May be null.</param>
+        /// <returns>The CloudEvent derived from the structured data.</returns>
+        public virtual CloudEvent DecodeStructuredModeMessage(byte[] data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes) =>
+            throw new NotImplementedException();
```

Comment:
Why aren't these implemented?

Reply:
Because they can't be - this code doesn't know how to decode messages :)
We could make the methods abstract instead... but if the point of having it as an abstract class rather than as an interface is to allow for further methods to be added without breaking binary compatibility, we should work out what we'd have done if we'd wanted to add these methods later. (Basically I'm looking for consistency - it's a little odd to have some methods abstract, and others just virtual. Maybe we can live with that, given that event formatters really should implement these...)

Reply:
Ah yeah - it seems like it would be correct to make this abstract in that case.

Reply:
Okay - I may do that in a trivial follow-up PR before anything more meaningful then.